app/controllers/application_controller.rb

```diff
@@ -5,6 +5,144 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "super_application_secret"
   end
 
+  get "/" do 
+    erb :index
+  end
+
+  get "/signup" do 
+    if logged_in?
+      redirect to "/tweets"
+    end
+    erb :"/users/create_user"
+  end
+
+  post "/signup" do 
+    if params["username"] == "" || params["email"] == "" || params["password"] == ""
+      redirect to "/signup"
+    else
+      @user = User.create(params)
+      session[:id] = @user.id # log in the user after they sign up
+      redirect to "/tweets"
+    end
+  end
+
+  get "/login" do 
+    if logged_in?
+      redirect to "/tweets"
+    else
+      erb :"/users/login"
+    end
+  end
+
+  post "/login" do
+    @user = User.find_by(username: params["username"])
+    if @user && @user.authenticate(params["password"])
+      session[:id] = @user.id
+      redirect to "/tweets"
+    else
+      redirect to "/login"
+    end
+  end
+
+  get "/logout" do 
+    if logged_in?
+      session.clear 
+      redirect to "/login"
+    else
+      redirect to "/"
+    end
+  end
+
+  get "/tweets" do # show all tweets
```

Comment:
This controller is overloaded. Consider separating your concerns by splitting some of these actions into separate controllers (i.e. a users controller and a tweets controller).